### ISSUE TITLE
RAdam Instability vs AdamW / Adam

### ISSUE THREAD

**Issue by danielhanchen** [LiyuanLucasLiu/RAdam#54]
Late to the party, but once again good work to you all @LiyuanLucasLiu !

So I was testing RAdam vs AdamW on simple linear models [ie Logistic Regression / Linear Regression]. Obviously for these small problems, using new methods is a bit overdoing it, but trying them on small problems [Sklearn datasets like Boston, MNIST, Wine] is also important :) 

After finding the best LR using the Learning Range Finder (which turns out to be the same LR for both [0.046]) + using gradient centralization + batch size = 16, with careful bias intialization (mean(y)), RAdam does seem more "stable" than AdamW.
![image](https://user-images.githubusercontent.com/23090290/95020314-1d47ee80-06b6-11eb-8d29-198d7a0a2803.png)

However, I noticed that if you do NOT standardize your data, RAdam's gradient diverges dramatically. The LR Range Test on NOT standardized data gave LR = 6.51e-05, which is super small. But, RAdam diverges.
![image](https://user-images.githubusercontent.com/23090290/95020513-6f3d4400-06b7-11eb-9024-0d791ee86ab4.png)

AdamW [lr = 1e-3] also has higher error when not standardized, however, the loss doesn't diverge a lot.
![image](https://user-images.githubusercontent.com/23090290/95020720-7fa1ee80-06b8-11eb-807a-07a0a803642c.png)

I also tried before (p < 5), to manually clip gradients by dividing by its norm. It's now much closer to AdamW.
![image](https://user-images.githubusercontent.com/23090290/95020774-e2938580-06b8-11eb-9426-befadead7493.png)

So my Q is: is this expected of RAdam to diverge if the dataset is not standardized? Should AdamW be used instead? Is it because of SGD + Momentum when (p < 5) that this divergement is seen?

**Comment by LiyuanLucasLiu** [comment]
Thanks for asking! 

I have noticed something similar to this. Setting `degenerated_to_sgd` to `False` resolved the issue for me (https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L7)

Specifically, I believe this issue is caused by that the gradient is too large in the begining of training. If the gradient is larger than one, then SGD would have larger step sizes than Adam/AdamW--adaptive optimizer scales learning rates with gradients' second momentom, if the second momentom is larger than one, directly removing it would leads to larger learning rate, which could potentially be over-large. Thus, in this situation, you can set `degenerated_to_sgd` to `False` and skip these updates.

**Comment by danielhanchen** [comment]
Thanks for the quick reply!! Appreciate it!

So I tried the suggestion where we set lr_t = -1* lr instead of lr_t = lr / (1 - beta1**t).
It does reduce the loss when (p < 5) somewhat, however the loss still diverges :(

The first image shows lr_t = -lr, [lr = 1e-3]. When p < 5, the loss is still very large [17 zeros]
![image](https://user-images.githubusercontent.com/23090290/95036929-1603fd80-0715-11eb-98b1-52168d63bccc.png)

The second image is when lr_t = lr / (1 - beta1**t). When p < 5, the loss is even larger than the first image [26 zeros]. So it partially solves the issue.
![image](https://user-images.githubusercontent.com/23090290/95036944-24521980-0715-11eb-8afb-8ca8bdc975da.png)

I'm currently trying to see how I can reduce the loss [maybe by change other params.] 

Anyways thanks once again for the swift reply!

**Comment by danielhanchen** [comment]
So I might have solved some of the issue.
When (p < 5), r becomes undefined. [As seen in the graph below]

![image](https://user-images.githubusercontent.com/23090290/95047571-0a72ff80-0732-11eb-826d-09ce797cb454.png)

So I just fit a line to do "analytic continuation" to kinda "derive" r when (p < 5). Not sure if this kinda reverts to back to using AdamW and variants. [r = 0.00523166 + 0.00419143*t] when (p < 5)

![image](https://user-images.githubusercontent.com/23090290/95047739-502fc800-0732-11eb-9a0b-2b45ec50f285.png)
![image](https://user-images.githubusercontent.com/23090290/95047932-a7ce3380-0732-11eb-932d-3642b5139b87.png)

You can see when t = 1, r = 0.00942309 or so. Until p = 5, then r uses the original equation.
You can see at least for now, RAdam actually trains like AdamW, but at a somewhat higher error.

![image](https://user-images.githubusercontent.com/23090290/95048109-fda2db80-0732-11eb-9a38-59ee4fd0bbd0.png)

Maybe one way to alleviate this issue is when p < 5, RAdam trains alongside AdamW. If RAdam goes haywire (like norm explodes), then RAdam reverts back to using AdamW.

[PS SGD with Momentum = 0.9 also diverges on unstandardized data].

**Comment by danielhanchen** [comment]
I also tried comparing AdamW / Adam's sensitivity to standard deviation of each feature.

As std(X)->inf, obvi the MSE would get larger. However, AdamW's loss doesn't diverge.
![image](https://user-images.githubusercontent.com/23090290/95061749-bc68f680-0747-11eb-878f-a1dc3924ecb2.png)

On the other hand due to RAdam using Momentum when (p<5) only, the MSE diverges after when std(X) exceeds 5 or so.
![image](https://user-images.githubusercontent.com/23090290/95061772-c4c13180-0747-11eb-8d6c-dffee66dce19.png)

Even at staggering large std(X)'s [like 100], AdamW successfully reduces the loss.
![image](https://user-images.githubusercontent.com/23090290/95061844-e02c3c80-0747-11eb-8b0d-48ba9daba1b2.png)

On the other hand, RAdam, well, diverges.
![image](https://user-images.githubusercontent.com/23090290/95061822-d73b6b00-0747-11eb-922b-ef94e1005c22.png)

So maybe from empirical tests, RAdam should default to AdamW if std(X) exceeds some number, say 5 or even 4? Otherwise, the loss can quickly diverge.

**Comment by danielhanchen** [comment]
Actually sorry on further investigation, it's not std(X) thats causing the issue. It's rather the max(abs(X)) thats the issue.
It looks like SGD with momentum diverges a lot when elements in X are large, whilst Adam can handle it due to the second moment essentially "factoring" out these large elements out.

From experiments, during the first batch, I found if
```norm(vtheta) > c * norm(mtheta)```
convergence for RAdam is much worse than Adam / AdamW.
c in my tests show 1.

What I find interesting is when standardization is on, in both RAdam and Adam(W), the norm of the second moment is ALWAYS a small number, whilst the norm of the first moment is much larger.

So for now, if I find after the first batch that `norm(vtheta) > norm(mtheta)`, I switch RAdam to use AdamW instead.

**Comment by LiyuanLucasLiu** [comment]
Hi, thanks for the discussion. For the implementation, setting lr to negative is not enough, I added a condition at https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L88 to skip updates when the learning rate is negative. Without adding this condition, the algorithm would do gradient ascent in the first few updates, which would cause serious problems.

**Comment by danielhanchen** [comment]
Ohhhhh oops I might have missed that! I'll read about it! Thanks again!

OHHH your right! Essentially if step_size < 0, you're skipping the first few SGD updates, then after the 5th iteration, the parameters get updated. Very interesting fix!

My bad I didn't notice that!

**Comment by danielhanchen** [comment]
So you're right! Though Glorot init kinda caused RAdam to have higher MSE than Adam. LSUV kinda solved this issue. Now, both RAdam and Adam even on weirdly large inputs seem to work.

![image](https://user-images.githubusercontent.com/23090290/95088959-4ed0c080-076f-11eb-8ef7-6f5171f8ce9d.png)

Anyways thanks once again for the quick replies!
For now, issues have been solved :) Thanks again!
